File: frontend/src/api/quotations.ts
import { apiClient } from './client';

export interface QuotationLineCreate {
  item_id?: string;
  item_name_snapshot: string;
  description?: string;
  hsn_sac_snapshot?: string;
  sku_snapshot?: string;
  quantity: number;
  unit_id?: string;
  unit_snapshot?: string;
  rate: number;
  discount_type?: string;
  discount_value?: number;
  gst_rate?: number;
}

export interface QuotationCreateRequest {
  quotation_type: string;
  tax_treatment: string;
  party_id: string;
  valid_until: string;
  place_of_supply: string;
  notes?: string;
  terms?: string;
  lines: QuotationLineCreate[];
}

export interface QuotationLineResponse {
  id: string;
  item_id: string | null;
  item_name_snapshot: string;
  description: string | null;
  hsn_sac_snapshot: string | null;
  sku_snapshot: string | null;
  quantity: number;
  converted_quantity: number;
  unit_id: string | null;
  unit_snapshot: string | null;
  rate: number;
  discount_type: string | null;
  discount_value: number;
  discount_amount: number;
  tax_treatment: string;
  gst_rate: number;
  taxable_value: number;
  cgst_amount: number;
  sgst_amount: number;
  igst_amount: number;
  cess_amount: number;
  line_total: number;
}

export interface QuotationResponse {
  id: string;
  quotation_number: string | null;
  quotation_type: string;
  tax_treatment: string;
  party_id: string;
  quotation_date: string;
  valid_until: string;
  status: string;
  revision: number;
  place_of_supply: string;
  subtotal: number;
  discount_total: number;
  taxable_total: number;
  cgst_total: number;
  sgst_total: number;
  igst_total: number;
  cess_total: number;
  round_off: number;
  grand_total: number;
  notes: string | null;
  terms: string | null;
  accepted_at: string | null;
  accepted_by: string | null;
  acceptance_method: string | null;
  fully_converted: boolean;
  created_at: string;
  lines: QuotationLineResponse[];
}

export const quotationsApi = {
  create: async (data: QuotationCreateRequest) => {
    const response = await apiClient.post<QuotationResponse>('/quotations', data);
    return response.data;
  },
  getAll: async (quotationType?: string) => {
    const params = new URLSearchParams();
    if (quotationType) params.append('quotation_type', quotationType);
    const response = await apiClient.get<{items: QuotationResponse[], total: number}>(`/quotations?${params.toString()}`);
    return response.data;
  },
  getById: async (id: string) => {
    const response = await apiClient.get<QuotationResponse>(`/quotations/${id}`);
    return response.data;
  },
  approve: async (id: string) => {
    const response = await apiClient.post<QuotationResponse>(`/quotations/${id}/approve`, {});
    return response.data;
  },
  accept: async (id: string, acceptanceMethod: string = "USER_ACCEPTED") => {
    const response = await apiClient.post<QuotationResponse>(`/quotations/${id}/accept`, {
      acceptance_method: acceptanceMethod
    });
    return response.data;
  }
};

File: backend/app/api/v1/quotations.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session
from app.core.database import get_db
from app.dependencies.auth import get_current_company, get_current_user
from app.schemas.common import ApiResponse
from app.schemas.quotation import (
    QuotationCreate, QuotationResponse, QuotationListResponse, QuotationAcceptRequest
)
from app.services.quotation_service import QuotationService

router = APIRouter(prefix="/quotations", tags=["Quotations"])

@router.post("/", response_model=ApiResponse[QuotationResponse])
def create_quotation(request: QuotationCreate, company = Depends(get_current_company), user = Depends(get_current_user), db: Session = Depends(get_db)):
    q = QuotationService.create_quotation(db, str(company.id), request.model_dump(), str(user.id))
    return ApiResponse(success=True, data=_quotation_to_response(q))

@router.get("/", response_model=ApiResponse[QuotationListResponse])
def list_quotations(quotation_type: str = Query(None), company = Depends(get_current_company), db: Session = Depends(get_db)):
    quotations = QuotationService.list_quotations(db, str(company.id), quotation_type)
    return ApiResponse(success=True, data=QuotationListResponse(
        items=[_quotation_to_response(q) for q in quotations],
        total=len(quotations)
    ))

@router.get("/{quotation_id}", response_model=ApiResponse[QuotationResponse])
def get_quotation(quotation_id: str, company = Depends(get_current_company), db: Session = Depends(get_db)):
    q = QuotationService.get_quotation(db, str(company.id), quotation_id)
    return ApiResponse(success=True, data=_quotation_to_response(q))

@router.post("/{quotation_id}/approve", response_model=ApiResponse[QuotationResponse])
def approve_quotation(quotation_id: str, company = Depends(get_current_company), db: Session = Depends(get_db)):
    q = QuotationService.approve_quotation(db, str(company.id), quotation_id)
    return ApiResponse(success=True, data=_quotation_to_response(q))

@router.post("/{quotation_id}/accept", response_model=ApiResponse[QuotationResponse])
def accept_quotation(quotation_id: str, request: QuotationAcceptRequest, company = Depends(get_current_company), user = Depends(get_current_user), db: Session = Depends(get_db)):
    q = QuotationService.accept_quotation(db, str(company.id), quotation_id, str(user.id), request.acceptance_method)
    return ApiResponse(success=True, data=_quotation_to_response(q))

def _quotation_to_response(q) -> QuotationResponse:
    lines = []
    for l in q.lines:
        lines.append({
            "id": l.id,
            "item_id": l.item_id,
            "item_name_snapshot": l.item_name_snapshot,
            "description": l.description,
            "hsn_sac_snapshot": l.hsn_sac_snapshot,
            "sku_snapshot": l.sku_snapshot,
            "quantity": float(l.quantity),
            "converted_quantity": float(l.converted_quantity),
            "unit_id": l.unit_id,
            "unit_snapshot": l.unit_snapshot,
            "rate": float(l.rate),
            "discount_type": l.discount_type,
            "discount_value": float(l.discount_value),
            "discount_amount": float(l.discount_amount),
            "tax_treatment": l.tax_treatment,
            "gst_rate": float(l.gst_rate),
            "taxable_value": float(l.taxable_value),
            "cgst_amount": float(l.cgst_amount),
            "sgst_amount": float(l.sgst_amount),
            "igst_amount": float(l.igst_amount),
            "cess_amount": float(l.cess_amount),
            "line_total": float(l.line_total)
        })
        
    return QuotationResponse(
        id=q.id,
        quotation_number=q.quotation_number,
        quotation_type=q.quotation_type.value,
        tax_treatment=q.tax_treatment,
        party_id=q.party_id,
        quotation_date=q.quotation_date,
        valid_until=q.valid_until,
        status=q.status.value,
        revision=q.revision,
        place_of_supply=q.place_of_supply,
        subtotal=float(q.subtotal),
        discount_total=float(q.discount_total),
        taxable_total=float(q.taxable_total),
        cgst_total=float(q.cgst_total),
        sgst_total=float(q.sgst_total),
        igst_total=float(q.igst_total),
        cess_total=float(q.cess_total),
        round_off=float(q.round_off),
        grand_total=float(q.grand_total),
        notes=q.notes,
        terms=q.terms,
        accepted_at=q.accepted_at,
        accepted_by=q.accepted_by,
        acceptance_method=q.acceptance_method,
        fully_converted=q.fully_converted,
        created_at=q.created_at,
        lines=lines
    )
